### DISCUSSION TITLE
Allow using SSH keys to sign commits

### DISCUSSION THREAD

**Discussion by blimmer** [community/community#7744] [Other Feature Feedback, Questions, & Ideas] [answered]
Git 2.34 now supports signing commits with SSH keys instead of GPG: https://github.blog/2021-11-15-highlights-from-git-2-34/#tidbits (see the first bullet point in the "tidbits" section).

It would be super cool to have this work with GitHub's existing "Verified Commit" tag, like it does with GPG keys.

**Comment by cgarrovillo** [comment]
gpg also often fails to sign commit data: 
https://stackoverflow.com/questions/41502146/git-gpg-onto-mac-osx-error-gpg-failed-to-sign-the-data/41506446
https://stackoverflow.com/questions/39494631/gpg-failed-to-sign-the-data-fatal-failed-to-write-commit-object-git-2-10-0?noredirect=1&lq=1
https://stackoverflow.com/questions/41052538/git-error-gpg-failed-to-sign-data?noredirect=1&lq=1

**Comment by mosajjal** [comment]
Saves a lot of us to carry two keys instead of one. Also looks like the SSH key signature method doesn't have the "type confusion" issue of GPG

**Reply by anarsen** [reply]
You _can_ use GPG keys for SSH auth too.

**Reply by craigfurman** [reply]
Your use case may vary, but personally I'm excited about this feature because signing commits (when mandated by company policy) is my only remaining usage of PGP, and I'm looking forward to putting it in the rear view mirror

**Reply by fazalmajid** [reply]
Yep. Apart from its general user-unfriendliness, it is also lined with security pitfalls:

https://latacora.micro.blog/2019/07/16/the-pgp-problem.html

**Comment by maxgoedjen** [comment]
+1. Here's a demo commit that's got a valid signature using an SSH key: https://github.com/maxgoedjen/ssh_signature_demo/commit/759243c407cbcc6b9d75724d59b9d3b49d8e917b, with a [key listed in my profile](https://github.com/maxgoedjen.keys)

Basically GitHub just doesn't read the type of keys yet, but would love for this to get added soon (I've been putting off adding GPG support to [Secretive](https://github.com/maxgoedjen/secretive) for ages, but this works great with it pretty much out of the box.

In terms of verifying keys – it seems like it'd be reasonable for it to show as "verified" with any keys that have push access to your account, although personally I'd love the ability to mark specific keys as being used for push access, singing verification, or both.

<img width="547" alt="image" src="https://user-images.githubusercontent.com/342665/142137860-dd8f8262-0d57-4c4a-83f8-194adf36fdce.png">
<img width="682" alt="image" src="https://user-images.githubusercontent.com/342665/142137872-c5cab9b9-108b-48bb-a674-6b866828cd4c.png">

**Reply by krubenok** [reply]
A necessary fix on the GitHub side for things like secretive to be useful for commit signing would be something similar to what's proposed here: isaacs/github#1099 (because the private key is tied to hardware that will eventually be retired)

Allow a key (ssh or gpg) to be "retired". Commits signed with with that key before the retirement date should be shown as valid with a sub-status that the key has since been retired, any commits made after the retirement would be invalid. This means the relatively ephemeral nature of secretive or other hardware backed keys is basically moot.

For additional use cases, when revoking a key present the following options:
- retire from date onwards (key was not compromised, just not using it anymore)
- revoke from date onwards (key was compromised, old commits are still valid)
- revoke and invalidate all previous uses (key was compromised, old commits are invalid because of uncertainty as to when the key was compromised)

**Reply by RoyTinker** [reply]
☝️ especially for organizations. An employee who left the org and later deletes her SSH keys should have no effect on the validity of prior signed commits made to org repos.

Sometimes I retire old keys because I got rid of an old computer. That shouldn't affect the validity of old signed commits, especially for my employer's repos.

**Reply by marvinroger** [reply]
In case you missed it: https://github.blog/changelog/2022-05-31-improved-verification-of-historic-git-commit-signatures/

**Reply by wiktor-k** [reply]
@marvinroger The post doesn't have anything about signing git commits with SSH (CTRL+F - "ssh" no hits) and nothing changed as of now (visit the link posted by Max: https://github.com/maxgoedjen/ssh_signature_demo/commit/759243c407cbcc6b9d75724d59b9d3b49d8e917b).

**Reply by marvinroger** [reply]
@wiktor-k yes, I am aware of that, however the post relates to the feature asked by @krubenok in the first comment of this thread:

> Allow a key (ssh or gpg) to be "retired". Commits signed with with that key before the retirement date should be shown as valid with a sub-status that the key has since been retired, any commits made after the retirement would be invalid. This means the relatively ephemeral nature of secretive or other hardware backed keys is basically moot.

**Reply by neo** [reply]
Thank you for sharing! With that update, do we still need to keep the expired public GPG keys in our account?

**Reply by jonaharagon** [reply]
It appears so, yes.

**Reply by krubenok** [reply]
Seems like some necessary pre-work to support SSH signing has shipped 🚢!

**Comment by jidicula** [comment]
Also, your SSH public keys are available at github.com/username.keys (unfortunately without their identifiers). This could potentially make it very easy for project maintainers to automatically create an `allowed-signers` file.

**Reply by jidicula** [reply]
I think we would need the key identifiers to be publicly visible for generating an `allowed-signers` file, though.

**Reply by unbeatable-101** [reply]
> I think we would need the key identifiers to be publicly visible for generating an `allowed-signers` file, though.

Yep, it doesn't work without them.

**Reply by madhead** [reply]
Isn't it a privacy threat? One may use a private corporate email as an identifier. Leaking it may disclose the fact of employment, which may be undesirable.

**Reply by hilli** [reply]
One could switch between ssh signing keys, if that leak is undesirable.

**Comment by farski** [comment]
I learned about this new functionality in 2.34 from the [GitHub blog](https://github.blog/2021-11-15-highlights-from-git-2-34/#tidbits), so I was quite surprised when I found out that GitHub doesn't actually support it yet.

**Comment by Menelion** [comment]
+1. Also, that will remove hassle caused by GPG for Windows users like me.

**Reply by unbeatable-101** [reply]
Don't worry, GPG is also a hassle to people not on windows 🥲

**Comment by davidstaheli** [comment, marked as answer]
Hi all. Thanks for this feedback because it helps us prioritize. The GitHub Repos team is also excited to see SSH commit signing support in Git. We're anxious for GitHub's commit validation to support it. Like what @farski said above, this is something that would have been great to support at the same time that Git began supporting it. We've recently expanded the GitHub Repos team and are still hiring to be more capable of that in the future. This issue is on our near-term backlog, but in full disclosure, we probably won't start working on it for 2-4 months. Commit security is an area that we want to give extra attention to during the next several months. @maxgoedjen, thanks for the good ideas about key management. If anyone has other suggestions on how repository or commit security can be improved, we'd love to hear them. We appreciate your perspective and knowing what is most important to you.

**Reply by dbrgn** [reply]
One way how the security with regards to SSH keys could be improved, would be if every SSH key could have a public name/reference in addition to the current (non-public) title. That key would then be included in the `https://github.com/<username>.keys` output. (Otherwise, if a user has multiple keys listed without any description, I have no idea whether I should trust some or all of them. Maybe I only want to trust the "work" key? Or only the "release signing" key?)

**Reply by politician** [reply]
@davidstaheli I am glad to hear that you want to focus on commit security. With many more organizations using GitHub Actions, Continuous Delivery, GitOps, etc. every day, git commits are increasingly becoming a **huge** attack vector. Commit signatures (as well as protected branches and required reviews) are a good way for security-conscious organizations to limit the attack surface.

Please, if you could pay extra attention to how keys are revoked, this is an important part where GitHub can add some value on top of what Git provides. My 2ç on [this issue](https://github.com/isaacs/github/issues/1099):

1. User revokes key/subkey on local machine
2. User uploads updated key containing revocation to GitHub
3. GitHub detects there is a revoked key/subkey that was used to sign commits previously and asks user if the key was compromised/lost and confirm a time and date from which to flag/unverify commits.
4. Before showing "unverified" status on GitHub, GitHub checks its cache to see if a commit hash was known as verified, so GitHub is now smarter than `git log --show-signature`

**Reply by taoso** [reply]
It finally works. Thank you 😄 
https://github.blog/changelog/2022-08-23-ssh-commit-verification-now-supported/

**Comment by LunNova** [comment]
When this is implemented, it would be nice to be able to add keys which are trusted for signing but not trusted for pushing, instead of having a single list.

**Reply by fproulx-boostsecurity** [reply]
^ yeah, so that we one has the option to keep a similar Threat Model as SSH keys vs GPG keys.

For instance, I like to be able to `git` clone, fetch, etc. without having to take my yubikey out of my pocket, but for signing commits, that's another thing.

**Comment by RoyTinker** [comment]
Our org (using GitHub enterprise) would love to introduce required commit signing to all protected branches, but has only done so for a few security-sensitive repos, since rolling out GPG commit signing across the org is so cumbersome.

Support for SSH commit signing in GitHub repos would be so helpful.

**Comment by fiadliel** [comment]
Some thoughts on what might be some useful features here:

1.  allow the option for GitHub-internal commits to be signed with an SSH key instead of GPG (also provide access to all previous public keys and valid periods of use). Configure per-repo and per-user/per-organization.
   This allows a single approach for validating all signatures in a repository.
1. allow a user or organization to provide an allowed-signers file (hardcoded, or potentially via a provided URL).
   This is especially for where organizations want full control over what keys are considered valid in their organization for signatures. e.g. an organization might only want signatures done via keys distributed on hardware tokens, and users should not be able to create a valid signature with a key generated any other way.
1. allow a user or organization to provide a key revocation list (hardcoded, or potentially via a provided URL).
1. potentially generate an allowed-signers sign automatically from uploaded public keys; this is what has been discussed so far in this issue.

**Comment by djmitche** [comment]
Hi there, a quick note on this topic -- the current behavior is
 * GPG signed -- nice "Verified" badge
 * Unsigned - no badge
 * SSH signed -- scary "Unverified" badge

This creates the perverse incentive to _disable_ signing of commits in order to avoid the false message to users that a repo contains malicious commits.  In other words, this is making software security strictly _worse_.

While I understand that it will take some time for GH to show a "Verified" badge, could the hopefully-simpler change to show no badge for SSH-signed commits be made sooner?

**Reply by Mikaela** [reply]
Using vigilant mode I would prefer something like "unrecognised"

I have been using SSH to sign my commits for about two months and Gitea has supported verifying them for a month

**Comment by xIGBClutchIx** [comment]
Any updates on this?

**Reply by davidstaheli** [reply]
I'm so happy to say that implementation of this feature has begun! 🎉

**Reply by MichaelDeBoey** [reply]
@davidstaheli Was this already implemented or is the team still working on this?

**Reply by davidstaheli** [reply]
Hi, @MichaelDeBoey! The team is still working on it, but I think they're almost done. I can get the latest details early next week.

**Reply by davidstaheli** [reply]
Update: this feature should be released by the end of June. 🚢

**Reply by MinisculeGirraffe** [reply]
Following up on this, as it's nearly the end of June. 

Seriously can't wait. It'd be so nice to kill my gpg agent.

**Reply by maxi-bee** [reply]
news?

**Reply by adamcstephens** [reply]
Please don't post a message asking about a status update. It notifies everybody following this thread but is just spam as it is not actually providing useful information to those of us following.

I understand it was stated that it would be out by the end of June, but we're only a few business days into the month and it is a slow holiday week in the US.

**Reply by jidicula** [reply]
All of GitHub is having a wellness holiday for this week (I read about it on Blind) so it definitely won't be shipping this week.

**Reply by davidstaheli** [reply]
Hi, all! I'm sorry to say that this slipped into July. The estimated release date is now July 20, but there's a decent chance it could ship before then. The pull request looks very close to done, if not already done. 🤞

**Reply by farski** [reply]
Maybe tomorrow???

**Reply by djmdjm** [reply]
Any update on when this is expected to launch?

**Reply by davidstaheli** [reply]
I apologize that this has taken longer than I anticipated. It has been released to GitHub's internal engineers for testing. Assuming all goes well, it should be released to everyone in the next two weeks (August 15).

**Reply by mhalano** [reply]
I think dogfooding is a wonderful practice

**Reply by fishtreesugar** [reply]
Is it delayed?

**Reply by mvdan** [reply]
Please don't ask for an update every two weeks. We would all like this featur sooner than later, but it's a very common time for holidays and travel right now. I would personally give it a rest until mid September at this point.

**Comment by jonaharagon** [comment]
I'm not sure if this has already been suggested, but I would appreciate a way to mark SSH and GPG keys as obsolete or revoked, such that commits made before they're marked obsolete remain Verified, but future commits made with the key are not, and the obsolete keys don't show up on https://github.com/JonahAragon.keys or https://github.com/JonahAragon.gpg.

It'd be especially handy once SSH signing here rolls out, so I can 'remove' my GPG keys from my account without unverifying every commit I've made in the past 2 years or so.

**Reply by juneidy** [reply]
I can see in the future where people think their private key is stolen or lost so they had to remove the public key for safety measure.

And yeah, you wouldn't want to have that unverified red sign everywhere.

**Reply by Mikaela** [reply]
👍🏻 This is possible with local `allowed_signers` file, which includes options `valid-after` and `valid-before` so I would expect GitHub to allow configuring those too.

* https://www.man7.org/linux/man-pages/man1/ssh-keygen.1.html#ALLOWED_SIGNERS

**Reply by marvinroger** [reply]
In case you missed it: https://github.blog/changelog/2022-05-31-improved-verification-of-historic-git-commit-signatures/

**Reply by fazalmajid** [reply]
That assumes there is a trusted commit time. GitHub can only know the time the commit was pushed to them and would have to use this timestamp and not the commit timestamp for the attestation.

**Reply by djmdjm** [reply]
git does provide direct support for revoking SSH signing keys, beyond the `valid-after|before` flags mentioned above:
```
     gpg.ssh.revocationFile
           Either a SSH KRL or a list of revoked public keys (without the
           principal prefix). See ssh-keygen(1) for details. If a public key
           is found in this file then it will always be treated as having
           trust level "never" and signatures will show as invalid.
```
Hopefully GitHub will surface a similar capability.

**Reply by davidstaheli** [reply]
Thanks for these great suggestions. We might get a first version released and then quickly follow up with some of these ideas. I'll try to share more details once the final scope is set. Though GitHub already recognizes when GPG keys are expired or revoked, I want to make sure we'll support the same for SSH, and if not in v1, how soon thereafter.

**Reply by abepolk** [reply]
Hi, I'm trying to understand the difference between signing commits with SSH and GPG. From a [GitHub docs article](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification), it seems that GitHub still doesn't recognize the revocation or expiry of SSH keys, but it does for GPG keys. The article makes it sound like this is because of a conceptual difference between the two technologies, but this thread makes it seem like GPG is supported in this way just because that's how GitHub was historically set up. Am I missing something here, or perhaps misinterpreting the docs? Thanks!

**Comment by krubenok** [comment]
https://github.blog/changelog/2022-08-23-ssh-commit-verification-now-supported/

🎊

**Reply by codemonkey85** [reply]
Am I missing something here? Looks like it works in theory but GitHub wants me to upload my SSH key, when I've already done that, and I've been using it for a while.

<details>
  <summary>Here are a couple of screenshots to show what I mean.</summary>

![image](https://user-images.githubusercontent.com/1414036/186290359-86d943d1-3e12-41dc-8a49-d20851b6502e.png)

![image](https://user-images.githubusercontent.com/1414036/186290367-d5fa4cfe-c74b-4f39-ae43-e57636f4864d.png)

</details>

**Reply by wyndon** [reply]
Did you read the link above?

> If you already use an SSH key to authenticate with GitHub, you can now upload the same or a different key pair’s public key to use it as a signing key. There is no limit to the number of signing keys you can add to your account.

**Reply by codemonkey85** [reply]
Ahhh, thanks @wyndon. I didn't read that far down, or realize I had to upload it again as a signing key.

**Comment by paulmillr** [comment]
**Update:** fixed it; if you intend to use ssh-signed commits, you must create an empty `allowedSignersFile` and point git config to it. See replies to the post.

> Today's release doesn't really work. GitHub worked a few months on a feature which is incomplete and is barely working.

> I've signed a commit, pushed to the repo, all cool. Now, someone clones my repo, does `git log --pretty='format:%GK'`, and what do they see?

>    `error: gpg.ssh.allowedSignersFile needs to be configured and exist for ssh signature verification`

> Same error for `git show`. So basically if I clone 15 repos, I will need to manually add every contributor who uses ssh signature, to the `allowedSignersFile`. How am I supposed to know their ssh keys? Why doesn't GH automatically create the file for every repo? 

> As a side note, `git log` works great with PGP keys, and displays all keys without a requirement of adding them to your keychain.

**Reply by mhalano** [reply]
> Today's release doesn't really work. GitHub worked a few months on a feature which is incomplete and is barely working.
> 
> I've signed a commit, pushed to the repo, all cool. Now, someone clones my repo, does `git log --pretty='format:%GK'`, and what do they see?
> 
> ```
> error: gpg.ssh.allowedSignersFile needs to be configured and exist for ssh signature verification
> ```
> 
> Same error for `git show`. So basically if I clone 15 repos, I will need to manually add every contributor who uses ssh signature, to the `allowedSignersFile`. How am I supposed to know their ssh keys? Why doesn't GH automatically create the file for every repo?
> 
> As a side note, `git log` works great with PGP keys, and displays all keys without a requirement of adding them to your keychain.

That's a limitation of SSH signing itself, not of GitHub implementation. You could get the SSH keys using the command: `curl https://github.com/<username>.keys`.
Maybe a file with all SSH keys generated automagically would be a good idea, but you must not add files to a repo without users' knowledge.

**Reply by jonaharagon** [reply]
This sounds like an issue with `git`, not GitHub. I'm not sure how this could be fixed since SSH keys don't include identity information like GPG keys do. GitHub's release today seems to work exactly as expected/promised (although it's not clear to me whether SSH signing keys can be revoked as I mentioned in https://github.com/orgs/community/discussions/7744#discussioncomment-2650109).

FWIW you can get people's keys by adding `.keys` to their profile URL, like https://github.com/jonaharagon.keys.

**Reply by paulmillr** [reply]
Without auto-generated `allowedSignersFile`, the feature is useless because it breaks `git log` and `git show`.

`curl <username>.keys` is obviously a bad idea: you cannot realistically do this for every contributor of every project you're interacting with. In other words, GH stands in a good position of auto-generating the file, because they have the full social graph.

**Reply by jrmann100** [reply]
This project might be relevant? https://github.com/frankywahl/allowedSignersFile

**Reply by Raniz85** [reply]
If `git log` and `git show` breaks that's a bug in Git _not_ GitHub.

Besides, the error message seems to indicate that you'll just have to create an empty file, configure Git to read it and then it'll work, but Git will show unverified commits - just as it does if you haven't imported the authors' GPG keys.

```
$ git verify-commit HEAD
gpg: Signature made Wed Aug 17 08:29:23 2022 UTC
gpg:                using RSA key 4AEE18F83AFDEB23
gpg: Can't check signature: No public key
```

You should _not_ trust commits just because Git can display their signature. For it to be worth anything you _must_ verify that the signature was made with a key that you trust.

`git log --pretty="format:%GK"` does _not_ verify the signatures, it just shows which key was used to sign. You want `git log --pretty="format:%G?"` which prints an `E` if the signature could not be verified.

After importing and trusting GitHub's web-flow GPG key and then trying again I now get this:
```
$ git verify-commit HEAD
gpg: Signature made Wed 17 Aug 2022 10:29:23 CEST
gpg:                using RSA key 4AEE18F83AFDEB23
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   2  signed:   7  trust: 0-, 0q, 0n, 0m, 0f, 2u
gpg: depth: 1  valid:   7  signed:   0  trust: 2-, 0q, 0n, 2m, 3f, 0u
gpg: Good signature from "GitHub (web-flow commit signing) <noreply@github.com>" [full]
```

**Reply by paulmillr** [reply]
> You should not trust commits just because Git can display their signature

Agree — though my concern was breakage of `git log` and `git show`, not semantics of `%G?` vs `%GK`. Creating an empty global `allowed_signers` fixed the issue.

Other users would still see the error if they don't create the file by default. This is a git issue and should be reported somewhere, but not GitHub's — empty file fixed it.

So, it's usable now.

**Reply by neo** [reply]
not sure if it's a bug, but I would like to point out that `https://github.com/<username>.keys` currently doesn't include the "signing keys" but only the "auth keys"

**Comment by fazalmajid** [comment]
Does the signature verification work with SSH CAs, which are supported in GitHub Enterprise, and if so is there an option to require the commit to be signed by a key that is certified by the repo’s org’s SSH key? Ideally I’d want the commits to be signed by an OpenSSH 8.2+ FIDO U2F ed25519-sk key certified by the CA key with a short-lived certificate so a stolen or lost Yubikey is not an issue.

**Reply by lpcalisi** [reply]
Hi @fazalmajid !  I've created a [discussion](https://github.com/orgs/community/discussions/55204) for this topic some time ago

**Comment by lucasresck** [comment]
Thank you very much for this long awaited SSH signing feature!

Would it be possible in the future to implement some kind of isolation of the SSH signing key to specific repositories only? I'm thinking more in a solution to the problem of an untrusted remote machine in which you manage a specific repository (usually through an SSH deploy key).

**Comment by rosmanov** [comment]
Thank you for this feature!

Do you have estimations when it will be available in GitHub Enterprise?

**Comment by Menelion** [comment]
Thank you for the feature! I get no errors but I don't see a Verified icon on my commits either. Am I missing something? Unfortunately cannot show it to you so far, it's a private repo.

**Comment by AnthonyEnr1quez** [comment]
Will the rest API endpoint for creating an ssh key be updated to accept a ssh key type (auth/signing)?

https://docs.github.com/en/rest/users/keys#create-a-public-ssh-key-for-the-authenticated-user

**Comment by mhalano** [comment]
Include the signing keys with auth keys under https://github.com/user.keys it would be nice. I don't think there will be security implications.

**Comment by djmdjm** [comment]
Are FIDO SSH keys supported? I added a signing key to my account and pushed a commit that was signed using a FIDO key, but it shows up as "unverified" in the Github UI: https://github.com/openssh/openssh-portable/commit/3a683a19fd116ea15ebf8aa13d02646cceb302a9

Verifying locally works fine:

```
$ git show --show-signature
commit 3a683a19fd116ea15ebf8aa13d02646cceb302a9 (HEAD -> master, origin/master, origin/HEAD)
Good "git" signature for djm@mindrot.org with ECDSA-SK key SHA256:JDubEpIXO0CfnM+iagFQce4MiJJDqTvgbBR80SuRwTQ
Author: Damien Miller <djm@mindrot.org>
Date:   Fri Aug 26 14:23:55 2022 +1000

    initial list of allowed signers

diff --git a/.git_allowed_signers b/.git_allowed_signers
...
```

**Reply by djmdjm** [reply]
I also notice that adding a SSH signing key to my account did not appear on my https://github.com/settings/security-log though subsequent actions do

**Reply by mhalano** [reply]
> I also notice that adding a SSH signing key to my account did not appear on my https://github.com/settings/security-log though subsequent actions do

I don't see my actions related to SSH signing either

**Reply by mhalano** [reply]
> Are FIDO SSH keys supported? I added a signing key to my account and pushed a commit that was signed using a FIDO key, but it shows up as "unverified" in the Github UI: [openssh/openssh-portable@3a683a1](https://github.com/openssh/openssh-portable/commit/3a683a19fd116ea15ebf8aa13d02646cceb302a9)
> 
> Verifying locally works fine:
> 
> ```
> $ git show --show-signature
> commit 3a683a19fd116ea15ebf8aa13d02646cceb302a9 (HEAD -> master, origin/master, origin/HEAD)
> Good "git" signature for djm@mindrot.org with ECDSA-SK key SHA256:JDubEpIXO0CfnM+iagFQce4MiJJDqTvgbBR80SuRwTQ
> Author: Damien Miller <djm@mindrot.org>
> Date:   Fri Aug 26 14:23:55 2022 +1000
> 
>     initial list of allowed signers
> 
> diff --git a/.git_allowed_signers b/.git_allowed_signers
> ...
> ```

It should work, but you must have your signing key added to GitHub.

**Reply by qbit** [reply]
Mine are working: https://github.com/qbit/xin/commit/0ae779fc60ef4c4891781f6720b6179ac7b02df8 (using FIDO) - I see a `public_key.verify` entry in my security log from when I added the pubkey.

**Comment by billmetangmo** [comment]
I am not able to sign commit with my ssh key. I got:
```
error: Load key "/tmp/.git_signing_key_tmpYlRvax": invalid format?
fatal: failed to write commit object
```
Does anybody has encountered this issue ?
I am using Linux on WSL . Don't know if it's related

**Reply by msladek** [reply]
No, it's not related to GitHub now allowing SSH key signing. It's an issue with the setup of your local machine.
Make sure you add your signing key defined in `user.signingkey` to the ssh-agent before committing.

**Reply by billmetangmo** [reply]
Oh you save my day :)  !

**Reply by notshriram** [reply]
Thanks @msladek!
For people facing the same issue, here's github docs on [how to add the ssh key to the agent](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#adding-your-ssh-key-to-the-ssh-agent).

**Comment by hilli** [comment]
I have updated the text on vigilant mode to reflect the new possibilities. PR here: https://github.com/github/github/pull/233947

**Comment by NickCao** [comment]
I noticed that from the view of users not logged in, commits signed with ssh are still marked as unverified, is this intended or a bug?

**Reply by NickCao** [reply]
Just checked again and now they are showed as verified.

**Comment by mhalano** [comment]
Can we have the sign SSH key on the https://github.com/user.keys file? I just could see the authentication SSH keys.
